File: python/metropolis.py
import numpy as np
import matplotlib as mpl
from scipy.stats import distributions
rnorm = np.random.normal
runif = np.random.rand
dgamma = distributions.gamma.logpdf
dnorm = distributions.norm.logpdf

# The data
age = np.array([13, 14, 14,12, 9, 15, 10, 14, 9, 14, 13, 12, 9, 10, 15, 11, 15, 11, 7, 13, 13, 10, 9, 6, 11, 15, 13, 10, 9, 9, 15, 14, 14, 10, 14, 11, 13, 14, 10])
price = np.array([2950, 2300, 3900, 2800, 5000, 2999, 3950, 2995, 4500, 2800, 1990, 3500, 5100, 3900, 2900, 4950, 2000, 3400, 8999, 4000, 2950, 3250, 3950, 4600, 4500, 1600, 3900, 4200, 6500, 3500, 2999, 2600, 3250, 2500, 2400, 3990, 4600, 450,4700])/1000.

n_params = 3


def calc_posterior(a, b, t, y=price, x=age):
    # Calculate joint posterior, given values for a, b and t

    # Priors on a,b
    logp = dnorm(a, 0, 10000) + dnorm(b, 0, 10000)
    # Prior on t
    logp += dgamma(t, 0.001, 0.001)
    # Calculate mu
    mu = a + b*x
    # Data likelihood
    logp += sum(dnorm(y, mu, t**-2))
    
    return logp


def sample(iterations=10000, tune_interval=100):
        
    # Initial proposal standard deviations
    prop_sd = [5, 5, 5]
    
    # Initialize trace for parameters
    trace = np.empty((iterations+1, n_params))

    # Set initial values
    trace[0] = 0,0,1
    # Initialize acceptance counts
    accepted = [0]*n_params

    # Calculate joint posterior for initial values
    current_log_prob = calc_posterior(*trace[0])
    
    for i in range(iterations):
    
        if not i%1000: print 'Iteration', i
    
        # Grab current parameter values
        current_params = trace[i]
    
        for j in range(n_params):
    
            # Get current value for parameter j
            p = trace[i].copy()
    
            # Propose new value
            if j==2:
                # Ensure tau is positive
                theta = np.exp(rnorm(np.log(current_params[j]), prop_sd[j]))
            else:
                theta = rnorm(current_params[j], prop_sd[j])
            
            # Insert new value 
            p[j] = theta
    
            # Calculate log posterior with proposed value
            proposed_log_prob = calc_posterior(*p)
    
            # Log-acceptance rate
            alpha = proposed_log_prob - current_log_prob
    
            # Sample a uniform random variate
            u = runif()
    
            # Test proposed value
            if np.log(u) < alpha:
                # Accept
                trace[i+1,j] = theta
                current_log_prob = proposed_log_prob
            else:
                # Reject
                trace[i+1,j] = trace[i,j]
    
            # Tune every <tune_interval> iterations
            if not (i+1) % tune_interval:
    
                # Calculate aceptance rate
                acceptance_rate = accepted[j]/tune_interval
    
                if acceptance_rate<0.2:
                    prop_sd[j] *= 0.9
                elif acceptance_rate>0.5:
                    prop_sd[j] *= 1.1
                else:
                    print 'Parameter %i is tuned!' % j
    
                accepted[j] = 0

    return trace

if __name__=='__main__':

    t = sample()
    print np.mean(t, 0)

File: julia/metropolis.jl
load("extras/Rmath.jl")

# The data
age   = [13., 14, 14,12, 9, 15, 10, 14, 9, 14, 13, 12, 9, 10, 15, 11,
         15, 11, 7, 13, 13, 10, 9, 6, 11, 15, 13, 10, 9, 9, 15, 14, 14,
         10, 14, 11, 13, 14, 10]
price = [2950, 2300, 3900, 2800, 5000, 2999, 3950, 2995, 4500, 2800,
         1990, 3500, 5100, 3900, 2900, 4950, 2000, 3400, 8999, 4000,
         2950, 3250, 3950, 4600, 4500, 1600, 3900, 4200, 6500, 3500,
         2999, 2600, 3250, 2500, 2400, 3990, 4600, 450,4700]/1000.

                        # dnorm log-density method for vector x and mu 
function dnorm(x::Vector{Float64}, mu::Vector{Float64}, sd::Float64, give_log::Bool)
     [dnorm(x[i], mu[i], sd, give_log) for i=1:length(mu)]
end
                                        # log joint posterior density
function calc_posterior(a::Number, b::Number, t::Number,
                        y::Vector{Float64}, x::Vector{Float64})
                                        # Priors on a,b
    logp  = dnorm(a, 0, 10000, true) + dnorm(b, 0, 10000, true)
                                        # Prior on t
    logp += dgamma(t, 0.001, 1000, true)
                                        # Calculate mu
    mu = a + b*x
                                        # Data likelihood
    logp + sum(dnorm(y, mu, 1/sqrt(t), true))
end
                                        # passing a vector [a, b, log(t)]
function calc_posterior(p::Vector{Float64}, y::Vector{Float64}, x::Vector{Float64})
    length(p) == 3 ? calc_posterior(p[1], p[2], exp(p[3]), y, x) :
                     error("argument p must be Array(Float64, (3,))")
end

function sample(iterations::Int, tune_interval::Int,
                y::Vector{Float64}, x::Vector{Float64})
                                        # Initial proposal standard deviations
    n_params = 3
    prop_sd  = fill(5., (n_params,))
                                        # Creat output array trace
    trace    = zeros(Float64, (iterations+1, n_params))
                                        # Initialize acceptance counts
    accepted = zeros(Int, (n_params,))
                                        # log posterior density at initial values
    current_log_prob = calc_posterior(reshape(trace[1,:], (n_params,)), y, x)

    for i=2:(iterations+1)              
        if !bool(i%1000); println("Iteration $i");end

        p = reshape(trace[i-1,:], (n_params,)) # initialize proposal vector
        for j=1:n_params
            p[j] += prop_sd[j] * randn() # proposal for parameter j
                                        # log posterior at proposed vector
            proposed_log_prob = calc_posterior(p, y, x)
                                        # Log-acceptance rate
            alpha = proposed_log_prob - current_log_prob
                                        # Test proposed value
            if log(rand()) < alpha
                current_log_prob = proposed_log_prob # Accept
                accepted[j] += 1
            else
                p[j] = trace[i-1,j]     # revert to previous value
            end
            
            ## Tune the standard deviations every <tune_interval> iterations
            if !bool(i % tune_interval)
                acceptance_rate = accepted[j]/tune_interval
                if acceptance_rate < 0.2
                    prop_sd[j] *= 0.9
                elseif acceptance_rate > 0.5
                    prop_sd[j] *= 1.1
                else
                    println("Parameter $j is tuned!")
                end
                accepted[j] = 0
            end
        end
        trace[i,:] = p
    end
    trace, prop_sd
end

